src/uuid.cr

```diff
@@ -107,16 +107,59 @@ struct UUID
     new(bytes, variant, version)
   end
 
+  def self.parse?(value : String, variant = nil, version = nil)
+    bytes = uninitialized UInt8[16]
+
+    case value.size
+    when 36 # Hyphenated
+      {8, 13, 18, 23}.each do |offset|
+        return if value[offset] != '-'
+      end
+      {0, 2, 4, 6, 9, 11, 14, 16, 19, 21, 24, 26, 28, 30, 32, 34}.each_with_index do |offset, i|
+        if hex = hex_pair_at? value, offset
+          bytes[i] = hex
+        else
+          return
+        end
+      end
+    when 32 # Hexstring
+      16.times do |i|
+        if hex = hex_pair_at? value, i * 2
+          bytes[i] = hex
+        else
+          return
+        end
+      end
+    when 45 # URN
+      return unless value.starts_with? "urn:uuid:"
+      {9, 11, 13, 15, 18, 20, 23, 25, 28, 30, 33, 35, 37, 39, 41, 43}.each_with_index do |offset, i|
+        if hex = hex_pair_at? value, offset
+          bytes[i] = hex
+        else
+          return
+        end
+      end
+    else
+      return
+    end
+
+    new(bytes, variant, version)
+  end
+
   # Raises `ArgumentError` if string `value` at index `i` doesn't contain hex
   # digit followed by another hex digit.
   private def self.hex_pair_at(value : String, i) : UInt8
+    hex_pair_at?(value, i) || raise ArgumentError.new [
+      "Invalid hex character at position #{i * 2} or #{i * 2 + 1}",
+      "expected '0' to '9', 'a' to 'f' or 'A' to 'F'",
+    ].join(", ")
```

Comment:
I realize this code already existed before. But I think we could take the opportunity to fix it. Creating an array and joining it is just wasteful. We can use a single string literal.
```suggestion
    hex_pair_at?(value, i) || raise ArgumentError.new "Invalid hex character at position #{i * 2} or #{i * 2 + 1}, expected '0' to '9', 'a' to 'f' or 'A' to 'F'"
```

Reply:
Agreed. I considered doing this, too. Was just trying to take as few liberties as needed, though. 😄

Reply:
I actually wonder if the slowdown when the UUID was invalid came from this bit of code and not from exceptions. @jgaskins Do you know if that's the case? Like, doing a benchmark with the raising variant, but with this change too.

Reply:
Ah, no, that wasn't the case.

Reply:
Yeah, tracing and unwinding the stack is magnitudes slower than allocating a couple of objects.